pulp_ansible/app/tasks/collections.py

```diff
@@ -608,13 +626,60 @@ async def _find_all_collections(self):
 
         await asyncio.gather(*tasks)
 
+    async def _can_be_optimized(self):
+        """Check last synced metadata time."""
+        root, api_version = await self._get_root_api(self.remote.url)
+        if api_version == 3:
+            downloader = self.remote.get_downloader(
+                url=root, silence_errors_for_response_status_codes={404}
+            )
+            try:
+                metadata = parse_metadata(await downloader.run())
+            except FileNotFoundError:
+                return False
+
+            sources = set()
+            if self.collection_info:
+                sources = {r.source for r in self.collection_info if r.source}
+            sources.add(self.remote.url)
+
+            published = metadata.get("published")
+            repo_metadata_time = self.repository.last_synced_metadata_time
+            repo_metadata_time = repo_metadata_time.isoformat() if repo_metadata_time else ""
+
+            # From serializers.DateTimeField.to_representation:
+            if repo_metadata_time.endswith("+00:00"):
+                repo_metadata_time = repo_metadata_time[:-6] + "Z"
+
+            if published == repo_metadata_time and len(sources) < 2:
+                msg = _(
+                    "no-op: {repo} did not change since last sync - {published}".format(
+                        repo=self.remote.url, published=published
+                    )
+                )
+                noop = ProgressReport(message=msg, code="noop")
+                noop.state = TASK_STATES.COMPLETED
+                noop.save()
+                return True
+
+            self.last_synced_metadata_time = published
+
+        return False
+
     async def run(self):
         """
         Build and emit `DeclarativeContent` from the ansible metadata.
         """
+        if self.repository.remote and self.remote.pk == self.repository.remote.pk:
+            can_be_optimized = await self._can_be_optimized()
+            if self.optimize and can_be_optimized:
+                return
```

Comment:
It would be good to add a logging statement at debug level here indicating that the no-op occurred also.